### PR TITLE
feat(trace mode): add retain-on-first-failure mode for traces

### PR DESCRIPTION
**Pull request by StevanFreeborn**
## Related Issue
Implements the changes suggested in #29531 

## Summary of Changes

- Updated `docs/src/test-api/class-testoptions.md` to include new `on-first-failure` mode including a description of the modes function
- Updated `packages/playwright/src/program.ts` to include `on-first-failure` as a valid trace mode.
- Updated `packages/playwright/src/worker/testTracing.ts` to include logic for capturing trace and abandoning trace when using `on-first-failure` mode
- Updated `packages/playwright/types/test.d.ts` and `utils/generate_types/overrides-test.d.ts` to include new mode.
- Updated `tests/page/pageTestApi.ts` to include new mode as valid option
- Updated `tests/playwright-test/playwright.artifacts.spec.ts` with test to validate trace is captured using the new mode.
- Updated `tests/playwright-test/playwright.trace.spec.ts` with tests to validate that a trace is captured under similar conditions as in the `retain-on-failure` mode, but also a test to validate that the trace is only captured on the first failure when retrying tests.

### CODE REVIEW

**Comment by StevanFreeborn** [comment]
Any advice on addressing the failing checks?

File: docs/src/test-api/class-testoptions.md [hunk 1]
```diff
@@ -546,8 +546,8 @@ export default defineConfig({
 
 ## property: TestOptions.trace
 * since: v1.10
-- type: <[Object]|[TraceMode]<"off"|"on"|"retain-on-failure"|"on-first-retry">>
-  - `mode` <[TraceMode]<"off"|"on"|"retain-on-failure"|"on-first-retry"|"on-all-retries">> Trace recording mode.
+- type: <[Object]|[TraceMode]<"off"|"on"|"retain-on-failure"|"on-first-retry"|"on-first-failure">>
```
[the remark this answers is not in the document]
Comment (StevanFreeborn) [reply docs/src/test-api/class-testoptions.md:549]: Ah yeah that makes sense. Will update.

File: packages/playwright/src/program.ts [hunk 2]
```diff
@@ -16,25 +16,24 @@
...
 import fs from 'fs';
 import path from 'path';
-import { Runner } from './runner/runner';
-import { stopProfiling, startProfiling, gracefullyProcessExitDoNotHang } from 'playwright-core/lib/utils';
-import { serializeError } from './util';
-import { showHTMLReport } from './reporters/html';
-import { createMergedReport } from './reporters/merge';
-import { loadConfigFromFileRestartIfNeeded, loadEmptyConfigForMergeReports } from './common/configLoader';
-import type { ConfigCLIOverrides } from './common/ipc';
+import { program } from 'playwright-core/lib/cli/program';
```
[the remark this answers is not in the document]
Comment (StevanFreeborn) [reply packages/playwright/src/program.ts:21]: Oh sorry about that I think that is an on save sort that is happening. Will correct.

File: packages/playwright/src/worker/testTracing.ts [hunk 3]
```diff
@@ -48,8 +47,21 @@ export class TestTracing {
     this._tracesDir = path.join(this._artifactsDir, 'traces');
   }
 
+  private shouldCaptureTrace() {
```
[the remark this answers is not in the document]
Comment (StevanFreeborn) [reply packages/playwright/src/worker/testTracing.ts:50]: Can do. Will update.

File: packages/playwright/src/worker/testTracing.ts [hunk 4]
```diff
@@ -59,9 +71,7 @@ export class TestTracing {
       this._options = { ...defaultTraceOptions, ...value, mode: (mode as string) === 'retry-with-trace' ? 'on-first-retry' : mode };
     }
 
-    let shouldCaptureTrace = this._options.mode === 'on' || this._options.mode === 'retain-on-failure' || (this._options.mode === 'on-first-retry' && this._testInfo.retry === 1) || (this._options.mode === 'on-all-retries' && this._testInfo.retry > 0);
-    shouldCaptureTrace = shouldCaptureTrace && !process.env.PW_TEST_DISABLE_TRACING;
-    if (!shouldCaptureTrace) {
+    if (this.shouldCaptureTrace() === false) {
```
[the remark this answers is not in the document]
Comment (StevanFreeborn) [reply packages/playwright/src/worker/testTracing.ts:74]: I can update to `!this._shouldCaptureTrace()`.

**Comment by pavelfeldman** [comment]
This looks superb!

**Review by pavelfeldman — APPROVED** [review]

**Comment by StevanFreeborn** [comment]
Thank you for the opportunity to contribute! I really enjoy working with Playwright. You all have built something really great.

**Review by 1234Brian — COMMENTED** [review]
Ok to me🫡

**Pull request merged**